add docker support

SnenxyTengoku opened the pull request:
currently adds a Dockerfile, provides a compose.yaml example and includes a github action to build `amd64` and `arm64` images and publishes them to ghcr.io

currently builds and start, ~~but in my case for some reason it will not actually deploy itself out of the container and is unreachable no matter what. i don't know if this is just me or if i'm screwing something up with my setup, so making this a draft for now~~ ask-js is required to publish to `0.0.0.0` instead of `127.0.0.1` for container setups in order to function.

dockerfile itself may be a mess / probably produce images a tad bit large for what it is (?) , could probably be improved but idk

SnenxyTengoku committed:
feat: add Dockerfile

SnenxyTengoku committed:
feat: add docker startup script

SnenxyTengoku committed:
chore: shorten dockerfile and add compose file

SnenxyTengoku committed:
docs: add instructions for docker compose

SnenxyTengoku committed:
fix: update build command in dockerfile

SnenxyTengoku committed:
chore: clean and expand compose file

ihateblueb commented:
initialized means once and then done i assume? `pnpm migration:apply` should probably be run on each start for a docker setup. `pnpm migrate` isnt a command here. also, for consistency, the askjs directory should probably be ask-js and scripts should be in a /scripts/ directory in the repo (ill do that one)

ihateblueb committed:
move docker start script and adjust workingdir in dockerfile

SnenxyTengoku committed:
Merge branch 'ihateblueb:main' into dockerfile

SnenxyTengoku committed:
fix(Dockerfile): do migration:apply every startup

SnenxyTengoku committed:
chore(compose.yaml): cleanup

SnenxyTengoku committed:
chore: remove docker script as its no longer necessary

SnenxyTengoku commented:
in that case i just removed the script since that's all it was doing, so its useless if `migration:apply` will be done every start

also some cleanup. i still can't access it outside of the container for some reason, so. waa

SnenxyTengoku committed:
feat: add github action workflow

SnenxyTengoku committed:
chore: update gitignore for cosign

SnenxyTengoku committed:
fix(workflow): use correct tags

SnenxyTengoku committed:
chore: remove compose file from repo, provide example in readme

SnenxyTengoku commented:
added a github action as well so images are built and released under the `dev` tag for everything that gets pushed to `main` in ghcr.io. these images require cosign to be signed and "correctly" build though, i can help guide through that or otherwise just remove it from the action

and also removed the compose file from the repo itself and just provided an example config in the readme instead since ^^^

ihateblueb commented on README.md:
looks neat, i was gonna remove "we" but you already got that

ihateblueb replied:
oh oh also compose.examle.yaml in root instead of this codeblock

ihateblueb replied:
done

ihateblueb commented on .github/workflows/docker-dev.yml:
`${{ github.repository_owner }}` exists so i wonder if theres a description one

ihateblueb commented on Dockerfile:
shouldnt the migration be put on the RUN?

SnenxyTengoku replied:
isn't migration done to prepare the database? if so then it has to be under cmd to do it on the users side

ihateblueb replied:
oh that makes sense

ihateblueb committed:
move compose example to it's own file rather than in the readme

SnenxyTengoku commented:
i'd say this is basically done but the issue with the whole "container starts up and claims its up but is completely inaccessible to outside world" issue still persists and i have no idea where to begin debugging that.

SnenxyTengoku committed:
Merge branch 'ihateblueb:main' into dockerfile

SnenxyTengoku committed:
docs: clean up docker instructions

SnenxyTengoku commented:
i need someone smarter than me to figure out why this is the case , since i've already tried it on my vps and locally on my laptop and they still don't work despite everything looking correct

i suppose if you'd like, you can merge this now (although ideally setup cosign on this repo first so the first action build doesn't explode) since everything builds and work other than that issue

SnenxyTengoku commented:
okay, so after Some Researching:tm: it appears that ask-js will need to support changing the listening host. since its internally using `127.0.0.1`, it won't actually work, it needs to be set to `0.0.0.0` within the container/config then can be accessed with `127.0.0.1` outside of it, which makes it usable.

i guess as a workaround if this isn't currently possible is to just throw like caddy in the same compose file and use it as a reverse proxy for it, but of course for Reasons i'd prefer that to be like. Last Resort or whatever

ihateblueb commented:
oh this is a pretty easy fix i can make, i'll get to that in a bit

ihateblueb committed:
host config option

ihateblueb commented:
i'll test this on my end, let me know if it works for you

ihateblueb committed:
fix tiny tiny typo

ihateblueb committed:
cosign pub key

ihateblueb committed:
bump version

ihateblueb commented:
thank you for the pr :3
